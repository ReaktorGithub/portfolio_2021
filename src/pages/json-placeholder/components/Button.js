import React from 'react';
import {useSelector} from "react-redux";

const style = {
    height: '1.7em',
    padding: '0 20px',
    outline: 'none',
    cursor: 'pointer',
    fontSize: '1em',
}

const Button = ({ text, actionBtn }) => {

    const { disable } = useSelector(state => state.main);

    return (
        <button
            onClick={actionBtn}
            disabled={disable}
            style={style}
        >
            { text }
        </button>
    );
};

export default Button;