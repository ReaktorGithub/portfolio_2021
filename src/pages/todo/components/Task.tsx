import React, {useContext, useEffect, useState} from 'react';
import styled from 'styled-components';
import Select from "./UI/Select";
import {TaskContext} from "../App";
import {ITaskContext, ITaskProps, Priority} from "../types";

const StyledTask = styled.div`
  display: flex;
  max-width: 500px;
  gap: 15px;
  align-items: start;
  padding: 10px;
  border: 1px solid grey;
  border-radius: 7px;
  background: white;

  p {
    margin: 2px 0 0 0;
    flex-grow: 1;
    font-size: 14px;
    cursor: pointer;
  }

  &:hover {
    box-shadow: inset 0 0 6px 3px rgba(0,0,0,0.2);
  }
  
  .cancel {
    margin-right: 10px;
  }
  
  textarea {
    resize: none;
    min-height: 50px;
    width: 100%;
  }
`

const Task: React.FC<ITaskProps> = ({ task }) => {

    const [selected, setSelected] = useState<Priority>('none');
    const [editMode, setEditMode] = useState<boolean>(false);
    const [temp, setTemp] = useState<string>('');
    const { addDone, addCancelled, confirmEdit, confirmMove } = useContext<ITaskContext>(TaskContext);

    useEffect(() => {
        setSelected(task.priority)
    }, [task.priority])

    function handleConfirmEdit() {
        if (temp === '') return;
        setEditMode(false);
        confirmEdit(task, temp);
    }

    function handleCancelEdit() {
        setEditMode(false);
    }

    function handleEdit(e: React.ChangeEvent<HTMLTextAreaElement>) {
        setTemp(e.target.value);
    }

    function handleOpenEdit() {
        setEditMode(true);
        setTemp(task.text);
    }

    function handleMove(e: React.ChangeEvent<HTMLSelectElement>) {
        setSelected(e.target.value as Priority);
        confirmMove(task, e.target.value as Priority);
    }

    return (
        <StyledTask className='Task'>
            <input
                type='checkbox'
                onClick={() => addDone(task)}
            />
            {
                editMode ?
                    <div>
                        <textarea
                            onChange={ handleEdit }
                            value={ temp }
                        >
                        </textarea>
                        <button className='cancel' onClick={ handleCancelEdit }>отмена</button>
                        <button data-cy='OK' onClick={ handleConfirmEdit }>ОК</button>
                    </div>
                    : <p onClick={ handleOpenEdit }>{task.text}</p>
            }
            <Select
                width='50px'
                height='1.3rem'
                selected={selected}
                handleChange={handleMove}
            />
            <button onClick={() => addCancelled(task)}>X</button>
        </StyledTask>
    );
};

export default Task;