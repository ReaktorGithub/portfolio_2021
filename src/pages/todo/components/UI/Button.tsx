import React from 'react';
import styled from 'styled-components';
import {IButtonProps} from "../../types";

const StyledButton = styled.button<{ color: 'red' | undefined}>`
    min-width: 120px;
    height: 1.8rem;
    outline: none;
    border: none;
    background: ${props => props.color === 'red' ? '#e54f4f' : '#f8d8bc'};
    cursor: pointer;
    border-radius: 5px;
    
    &:hover {
        box-shadow: inset -2px -2px 4px 2px rgba(0,0,0,0.2);
    }
`

const Button: React.FC<IButtonProps> = ({ text, color }) => {
    return (
        <StyledButton color={color}>
            { text }
        </StyledButton>
    );
};

export default React.memo(Button);