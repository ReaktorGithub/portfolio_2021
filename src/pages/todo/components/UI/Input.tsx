import React from 'react';
import styled from 'styled-components';
import {IInputProps} from "../../types";

const StyledInput = styled.input`
    width: 100%;
    max-width: 250px;
    height: 1.8rem;
    outline: none;
    
    &:focus {
        box-shadow: 0 0 4px 2px rgba(0,0,0,0.1);
    }
`

const Input: React.FC<IInputProps> = ({ placeholder, handleInput, value }) => {

    return (
        <StyledInput
            type='text'
            placeholder={placeholder}
            onChange={handleInput}
            required
            value={value}
        />
    );
};

export default React.memo(Input);