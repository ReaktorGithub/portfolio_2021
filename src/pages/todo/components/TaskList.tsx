import React, {useEffect, useState} from 'react';
import Task from "./Task";
import styled from 'styled-components';
import {ITaskListProps, Priority} from "../types";

const StyledTaskList = styled.div<{type: Priority}>`
  width: 100%;
  height: 100%;
  border-radius: 7px;
  padding: 20px 10px;
  background: ${props => {
      switch (props.type) {
          case 'priority1': return '#fddbdb';
          case 'priority2': return '#ffd6b7';
          case 'priority3': return '#addbff';
          case 'priority4': return '#d2ffc2';
          default: return '#d4d4d4';
      }
}};
  .heading {
    width: 60%;
    min-width: 240px;
    padding: 10px;
    background: ${props => {
    switch (props.type) {
        case 'priority1': return '#ff9797';
        case 'priority2': return '#ffb77e';
        case 'priority3': return '#83d4ff';
        case 'priority4': return '#a2ff8a';
        default: return '#959595';
    }
}};
    box-shadow: 3px 3px 6px 2px rgba(0,0,0,0.1);
    margin: 0 0 20px -20px;
    
    h2 {
        font-family: "Russo One", sans-serif;
        font-size: 18px;
        margin: 0;
    }
  }
  
  .list {
      display: flex;
      flex-direction: column;
      gap: 10px;
  }
  
  .empty {
    margin: 0 0 0 20px;
    color: #8a8a8a;
    
  }
`

const headers = {
    none: 'Не сортировано',
    priority1: 'Важные срочные',
    priority2: 'Важные несрочные',
    priority3: 'Неважные срочные',
    priority4: 'Неважные несрочные',
}

const TaskList: React.FC<ITaskListProps> = ({ list, type }) => {

    const [empty, setEmpty] = useState<boolean>(true);

    useEffect(() => {
        for (let i = 0; i < list.length; i++) {
            if (list[i].priority === type) {
                setEmpty(false);
                break;
            }
            setEmpty(true);
        }
    }, [list, type])

    return (
        <StyledTaskList
            type={type}
            data-cy='TodoList'
        >
            <div className='heading'>
                <h2>{ headers[type] ? headers[type] : headers.none }</h2>
            </div>
            <div className='list'>
                {
                    list.map(item => {
                        if (item.priority === type) {
                            return <Task
                                key={item.id}
                                task={item}
                            />
                        }
                    })
                }
                {
                    (empty || !list.length) && <p className='empty'>пусто</p>
                }
            </div>
        </StyledTaskList>
    );
};

export default TaskList;
