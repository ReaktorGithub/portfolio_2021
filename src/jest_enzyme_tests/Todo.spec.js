import Button from "../pages/todo/components/UI/Button";
import {configure, mount, render, shallow} from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from "enzyme-to-json";
import Input from "../pages/todo/components/UI/Input";
import TaskList from "../pages/todo/components/TaskList";
import Task from "../pages/todo/components/Task";

configure({ adapter: new Adapter() });

describe('тестирование Todo', function () {

    describe('Тест Todo: Button', () => {

        it('отрисовка Button', () => {
            const comp = mount(<Button/>);
            const wrapper = comp.find('button');
            console.log(wrapper.html());
            expect(wrapper.length).toBe(1);
        })

        // снапшоты

        // способ 1
        it("renders correctly", () => {
            const tree = shallow(<Button />);
            expect(toJson(tree)).toMatchSnapshot();
        });

        // способ 2, без toJson и с render
        const snapshot = (text, color) => {
            const wrapper = render(<Button text={text} color={color} />);
            it(`text = ${text}, color = ${color}`, () => {
                expect(wrapper).toMatchSnapshot();
            });
        }

        snapshot('Я снова кнопка');
        snapshot('Я красная кнопка', 'red');
    });

    describe('Тест Todo: Input', function () {

        it('отрисовка Input', () => {
            const input = mount(<Input/>);
            const wrapper = input.find('input');
            expect(wrapper.length).toBe(1);
            console.log(wrapper.html());
        })

        it('снапшот Input', () => {
            const input = shallow(<Input/>);
            expect(toJson(input)).toMatchSnapshot();
        })

        const props = {
            placeholder: 'Введите пароль',
            handleInput: () => {},
            value: 'пароль',
        }

        it('инпут должен принять пропсы', () => {
            const wrapper = shallow(<Input {...props}/>);
            expect(wrapper.props().placeholder).toEqual('Введите пароль');
            expect(wrapper.props().value).toEqual('пароль');
            console.log(wrapper.html());
        })
    });

    describe('Тест Todo: TaskList', () => {

        const type = 'none';
        const list = [
            {
                id: 1,
                text: 'Бегать',
                priority: 'none',
            },
            {
                id: 2,
                text: 'Позвонить на работу',
                priority: 'priority1',
            },
        ]

        it('снапшот отрисовки списка', () => {
            const wrapper = mount(<TaskList type={type} list={list}/>);
            expect(wrapper).toMatchSnapshot();
            console.log(wrapper.html());
        })

        it("разметка должна содержать p > Бегать", () => {
            const wrapper = mount(<TaskList type={type} list={list} />);
            const value = wrapper.find("p").text();
            console.log(value);
            expect(value).toEqual("Бегать");
        });
    })

    describe('Тест Todo: Task', () => {

        let component;
        beforeEach(() => {
            component = shallow(<Task task={ {
                id: 1,
                text: 'Бегать',
                priority: 'none',
            } }/>);
        })

        it('снапшот Task', () => {
            expect(component).toMatchSnapshot();
        })

        it('рендер Task', () => {
            const wrapper = component.find('.Task');
            expect(wrapper.length).toBe(1);
        })
    })
});

