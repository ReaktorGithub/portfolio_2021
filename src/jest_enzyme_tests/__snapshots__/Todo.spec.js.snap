// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тест Todo: Button renders correctly 1`] = `<styled.button />`;

exports[`Тест Todo: Button text = Я красная кнопка, color = red 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "sc-bdnxRM dZoUtS",
      "color": "red",
    },
    "children": Array [
      Node {
        "data": "Я красная кнопка",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "color": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "color": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Тест Todo: Button text = Я снова кнопка, color = undefined 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "sc-bdnxRM bOCwnx",
    },
    "children": Array [
      Node {
        "data": "Я снова кнопка",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Тест Todo: Input снапшот Input 1`] = `
<styled.input
  required={true}
  type="text"
/>
`;

exports[`Тест Todo: Task снапшот Task 1`] = `ShallowWrapper {}`;

exports[`Тест Todo: TaskList снапшот отрисовки списка 1`] = `ReactWrapper {}`;

exports[`тестирование Todo Тест Todo: Button renders correctly 1`] = `<styled.button />`;

exports[`тестирование Todo Тест Todo: Button text = Я красная кнопка, color = red 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "sc-bdfBQB czLnHZ",
      "color": "red",
    },
    "children": Array [
      Node {
        "data": "Я красная кнопка",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "color": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "color": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`тестирование Todo Тест Todo: Button text = Я снова кнопка, color = undefined 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "sc-bdfBQB kWsaK",
    },
    "children": Array [
      Node {
        "data": "Я снова кнопка",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`тестирование Todo Тест Todo: Input снапшот Input 1`] = `
<styled.input
  required={true}
  type="text"
/>
`;

exports[`тестирование Todo Тест Todo: Task снапшот Task 1`] = `ShallowWrapper {}`;

exports[`тестирование Todo Тест Todo: TaskList снапшот отрисовки списка 1`] = `ReactWrapper {}`;
